File: backend/app/tool_catalog.py
"""Werkzeug-Katalog: verbindet die Kacheln der Oberfläche mit ihren Endpunkten.

Der Katalog ist die einzige Quelle dafür, welche Werkzeuge es gibt und welche
API-Pfade zu einem Werkzeug gehören. Damit kann der Administrator einzelne
Werkzeuge auf angemeldete Nutzer beschränken (siehe app/tool_access.py) und die
Beschränkung wird serverseitig durchgesetzt — nicht nur in der Oberfläche.

Bewusst NICHT im Katalog: gemeinsam genutzte Lese-/Vorschau-Endpunkte
(Thumbnails, Metadaten lesen, Formularfelder lesen, TOC lesen, Bilder
auflisten). Sie gehören zu keinem einzelnen Werkzeug und werden von mehreren
Ansichten für die Vorschau gebraucht; gesperrt wird die eigentliche Aktion.

Pfade mit abschließendem "/" gelten als Präfix (z.B. parametrisierte Routen).
"""

from dataclasses import dataclass


@dataclass(frozen=True)
class ToolEntry:
    id: str
    label: str
    group: str
    paths: tuple[str, ...]


_PDF = "PDF"
_OFFICE = "Word & Excel"

TOOL_CATALOG: tuple[ToolEntry, ...] = (
    ToolEntry("textEditor", "Text bearbeiten", _PDF, ("/api/pdf-extras/edit-text-blocks",)),
    ToolEntry("annotate", "Anmerkungen", _PDF, ("/api/pdf-editor/annotations",)),
    ToolEntry("merge", "Zusammenführen", _PDF, ("/api/pdf-tools/merge",)),
    ToolEntry("split", "Teilen", _PDF, ("/api/pdf-tools/split",)),
    ToolEntry("rotate", "Rotieren", _PDF, ("/api/pdf-tools/rotate",)),
    ToolEntry("toImages", "PDF → Bilder", _PDF, ("/api/pdf-tools/to-images",)),
    ToolEntry("compress", "Komprimieren", _PDF, ("/api/pdf-tools/compress",)),
    ToolEntry("watermark", "Wasserzeichen", _PDF, ("/api/pdf-tools/watermark",)),
    ToolEntry("imagesToPdf", "Bilder → PDF", _PDF, ("/api/pdf-tools/images-to-pdf",)),
    ToolEntry("pageEditor", "Seiten ordnen", _PDF, ("/api/pdf-tools/page-operations",)),
    ToolEntry("insertBlank", "Leere Seiten", _PDF, ("/api/pdf-more/insert-blank",)),
    ToolEntry(
        "pdfCompare",
        "PDF Vergleich",
        _PDF,
        ("/api/pdf-editor/compare", "/api/pdf-extras/jobs/compare"),
    ),
    ToolEntry("ocr", "OCR", _PDF, ("/api/pdf-editor/ocr", "/api/pdf-extras/jobs/ocr")),
    ToolEntry(
        "protect",
        "Schützen / Entsperren",
        _PDF,
        ("/api/pdf-tools/protect", "/api/pdf-tools/unlock"),
    ),
    ToolEntry("permissions", "Rechteverwaltung", _PDF, ("/api/pdf-more/permissions",)),
    ToolEntry("clean", "Bereinigen", _PDF, ("/api/pdf-more/clean",)),
    ToolEntry("pdfToWord", "PDF → Word", _PDF, ("/api/pdf-convert/to-word",)),
    ToolEntry(
        "pdfToExcel",
        "PDF → Excel",
        _PDF,
        ("/api/pdf-convert/to-excel", "/api/pdf-convert/to-excel/preview"),
    ),
    ToolEntry(
        "pdfExport",
        "PDF exportieren",
        _PDF,
        ("/api/pdf-more/export/", "/api/pdf-more/to-csv"),
    ),
    ToolEntry("metadata", "Metadaten", _PDF, ("/api/pdf-tools/metadata",)),
    ToolEntry("toText", "PDF → Text", _PDF, ("/api/pdf-extras/to-text",)),
    ToolEntry(
        "searchReplace",
        "Suchen & Ersetzen",
        _PDF,
        (
            "/api/pdf-editor/text/replace",
            "/api/pdf-editor/text/insert",
            "/api/pdf-editor/text/search",
        ),
    ),
    ToolEntry(
        "redact",
        "Schwärzen",
        _PDF,
        ("/api/pdf-extras/redact", "/api/pdf-extras/redact-patterns"),
    ),
    ToolEntry("pageNumbers", "Seitenzahlen", _PDF, ("/api/pdf-editor/page-numbers",)),
    ToolEntry("headerFooter", "Kopf-/Fußzeile", _PDF, ("/api/pdf-editor/header-footer",)),
    ToolEntry("bates", "Bates-Nummern", _PDF, ("/api/pdf-extras/bates",)),
    ToolEntry("pruefakte", "Prüfakte", _PDF, ("/api/pdf-more/pruefakte",)),
    ToolEntry("formDesigner", "Formular-Designer", _PDF, ("/api/pdf-extras/form/create",)),
    ToolEntry(
        "formFill",
        "Formular ausfüllen",
        _PDF,
        (
            "/api/pdf-extras/form/fill",
            "/api/pdf-extras/form/fill-csv",
            "/api/pdf-extras/form/export-csv",
        ),
    ),
    ToolEntry("signImage", "Unterschrift", _PDF, ("/api/pdf-extras/sign-image",)),
    ToolEntry("imageReplace", "Bild ersetzen", _PDF, ("/api/pdf-more/images/replace",)),
    ToolEntry("pdfa", "PDF/A", _PDF, ("/api/pdf-extras/to-pdfa",)),
    ToolEntry("flatten", "Einbrennen", _PDF, ("/api/pdf-editor/flatten",)),
    ToolEntry("signDigital", "Digital signieren", _PDF, ("/api/pdf-extras/sign-digital",)),
    ToolEntry(
        "verifySignatures", "Signatur prüfen", _PDF, ("/api/pdf-more/verify-signatures",)
    ),
    ToolEntry(
        "scanOptimize",
        "Scan optimieren",
        _PDF,
        ("/api/pdf-extras/scan-optimize", "/api/pdf-extras/jobs/scan-optimize"),
    ),
    ToolEntry("qualityCheck", "Qualitätsprüfung", _PDF, ("/api/pdf-more/quality-check",)),
    ToolEntry("batch", "Batch", _PDF, ("/api/pdf-extras/batch",)),
    ToolEntry(
        "tocEditor",
        "Lesezeichen",
        _PDF,
        ("/api/pdf-editor/toc/write", "/api/pdf-editor/toc/auto-detect"),
    ),
    ToolEntry("aiAssistant", "KI-Assistent", _PDF, ("/api/pdf-more/ai/",)),
    ToolEntry("officeToPdf", "Office → PDF", _OFFICE, ("/api/pdf-more/office-to-pdf",)),
    ToolEntry("wordToPdf", "Word → PDF", _OFFICE, ("/api/pdf-tools/word-to-pdf",)),
    ToolEntry("wordMerge", "Word verbinden", _OFFICE, ("/api/pdf-tools/word-merge",)),
    ToolEntry("wordDiff", "Word Vergleich", _OFFICE, ("/api/pdf-tools/word-diff",)),
    ToolEntry("wordMeta", "Word Metadaten", _OFFICE, ("/api/pdf-tools/word-metadata",)),
    ToolEntry("excelMeta", "Excel Metadaten", _OFFICE, ("/api/pdf-tools/excel-metadata",)),
)

TOOL_IDS: frozenset[str] = frozenset(t.id for t in TOOL_CATALOG)

_EXACT: dict[str, str] = {
    path: tool.id for tool in TOOL_CATALOG for path in tool.paths if not path.endswith("/")
}
_PREFIX: tuple[tuple[str, str], ...] = tuple(
    (path, tool.id) for tool in TOOL_CATALOG for path in tool.paths if path.endswith("/")
)


def resolve_tool(path: str) -> str | None:
    """Werkzeug-ID zu einem Request-Pfad — None für gemeinsame Endpunkte."""
    tool = _EXACT.get(path)
    if tool:
        return tool
    for prefix, tool_id in _PREFIX:
        if path.startswith(prefix):
            return tool_id
    return None

File: frontend/src/lib/toolCatalog.ts
/**
 * Zentraler Katalog aller in der Startansicht angebotenen Werkzeuge.
 *
 * Suche, Breadcrumbs, Fokus-Header und Komponenten-Mapping verwenden dieselben
 * Metadaten. `ToolId` wird direkt aus diesem unveränderlichen Katalog abgeleitet;
 * fehlende oder falsch geschriebene Komponenten-IDs fallen damit bereits bei
 * der TypeScript-Prüfung auf.
 */

export interface ToolDefinition {
  id: string
  label: string
  description: string
  icon: string
  requiresAi?: boolean
}

export interface ToolGroupDefinition {
  key: string
  title: string
  icon: string
  tools: readonly ToolDefinition[]
}

export const TOOL_GROUPS = [
  {
    key: 'bearbeiten', title: 'Bearbeiten & Anmerken', icon: '✏️',
    tools: [
      { id: 'textEditor', label: 'Text bearbeiten', description: 'WYSIWYG direkt im Dokument', icon: '✏️' },
      { id: 'annotate', label: 'Anmerkungen', description: 'Markieren, stempeln und zeichnen', icon: '🖍️' },
      { id: 'searchReplace', label: 'Suchen & Ersetzen', description: 'Textstellen gezielt austauschen', icon: '🔁' },
      { id: 'signImage', label: 'Unterschrift', description: 'Bildsignatur präzise einfügen', icon: '✍️' },
      { id: 'imageReplace', label: 'Bild ersetzen', description: 'Logos und Grafiken austauschen', icon: '🖌️' },
      { id: 'formDesigner', label: 'Formular-Designer', description: 'Interaktive Felder neu anlegen', icon: '🧩' },
      { id: 'formFill', label: 'Formular ausfüllen', description: 'Vorhandene AcroForm-Felder bearbeiten', icon: '📋' },
      { id: 'tocEditor', label: 'Lesezeichen', description: 'Inhaltsstruktur und Sprungziele pflegen', icon: '🔖' },
    ],
  },
  {
    key: 'seiten', title: 'Seiten & Struktur', icon: '📑',
    tools: [
      { id: 'merge', label: 'Zusammenführen', description: 'Mehrere PDFs in Reihenfolge verbinden', icon: '📎' },
      { id: 'split', label: 'Teilen', description: 'PDF in definierte Bereiche aufteilen', icon: '✂️' },
      { id: 'rotate', label: 'Rotieren', description: 'Einzelne oder alle Seiten drehen', icon: '🔄' },
      { id: 'pageEditor', label: 'Seiten ordnen', description: 'Seiten verschieben oder entfernen', icon: '📑' },
      { id: 'insertBlank', label: 'Leere Seiten', description: 'Zusätzliche Seiten einfügen', icon: '➕' },
      { id: 'pdfCompare', label: 'PDF-Vergleich', description: 'Zwei Dokumentstände vergleichen', icon: '⚖️' },
    ],
  },
  {
    key: 'umwandeln', title: 'Umwandeln & Exportieren', icon: '🔁',
    tools: [
      { id: 'pdfToWord', label: 'PDF → Word', description: 'In eine bearbeitbare DOCX-Datei exportieren', icon: '📝' },
      { id: 'pdfToExcel', label: 'PDF → Excel', description: 'Tabellen als Arbeitsmappe extrahieren', icon: '📊' },
      { id: 'toImages', label: 'PDF → Bilder', description: 'Seiten als PNG oder JPEG ausgeben', icon: '🖼️' },
      { id: 'imagesToPdf', label: 'Bilder → PDF', description: 'Bilddateien zu einem PDF bündeln', icon: '📄' },
      { id: 'toText', label: 'PDF → Text', description: 'Reinen Dokumenttext extrahieren', icon: '📃' },
      { id: 'pdfExport', label: 'PDF exportieren', description: 'Markdown, HTML, JSON oder CSV erzeugen', icon: '📤' },
      { id: 'ocr', label: 'OCR', description: 'Scans durchsuchbar machen', icon: '🔍' },
      { id: 'pdfa', label: 'PDF/A', description: 'Dokument in ein Archivformat überführen', icon: '🗄️' },
    ],
  },
  {
    key: 'schuetzen', title: 'Schützen & Signieren', icon: '🔏',
    tools: [
      { id: 'protect', label: 'Schützen / Entsperren', description: 'Passwortschutz setzen oder entfernen', icon: '🔒' },
      { id: 'permissions', label: 'Rechteverwaltung', description: 'Drucken und Kopieren einschränken', icon: '🛡️' },
      { id: 'redact', label: 'Schwärzen', description: 'Sensible Inhalte unwiderruflich entfernen', icon: '⬛' },
      { id: 'clean', label: 'Bereinigen', description: 'Metadaten und versteckte Inhalte entfernen', icon: '🧽' },
      { id: 'flatten', label: 'Einbrennen', description: 'Anmerkungen dauerhaft fixieren', icon: '🔥' },
      { id: 'signDigital', label: 'Digital signieren', description: 'Zertifikatsbasierte PAdES-Signatur', icon: '🔏' },
      { id: 'verifySignatures', label: 'Signatur prüfen', description: 'Integrität und Unterzeichner kontrollieren', icon: '✅' },
    ],
  },
  {
    key: 'stempel', title: 'Stempel & Nummern', icon: '🏷️',
    tools: [
      { id: 'watermark', label: 'Wasserzeichen', description: 'Text oder Bild dezent überlagern', icon: '💧' },
      { id: 'pageNumbers', label: 'Seitenzahlen', description: 'Flexible Nummerierungen ergänzen', icon: '🔢' },
      { id: 'headerFooter', label: 'Kopf- / Fußzeile', description: 'Wiederkehrende Texte positionieren', icon: '📰' },
      { id: 'bates', label: 'Bates-Nummern', description: 'Dokumente revisionssicher kennzeichnen', icon: '🏷️' },
      { id: 'pruefakte', label: 'Prüfakte', description: 'Akte mit Lesezeichen und Bates erzeugen', icon: '🗂️' },
    ],
  },
  {
    key: 'pruefen', title: 'Prüfen & Optimieren', icon: '🩺',
    tools: [
      { id: 'compress', label: 'Komprimieren', description: 'Dateigröße kontrolliert reduzieren', icon: '🗜️' },
      { id: 'scanOptimize', label: 'Scan optimieren', description: 'Geraderücken, säubern und erkennen', icon: '🧹' },
      { id: 'qualityCheck', label: 'Qualitätsprüfung', description: 'Leere oder doppelte Seiten finden', icon: '🩺' },
      { id: 'metadata', label: 'Metadaten', description: 'Titel, Autor und Datumswerte bearbeiten', icon: 'ℹ️' },
      { id: 'batch', label: 'Batch', description: 'Viele Dateien in einem Durchlauf verarbeiten', icon: '📦' },
      { id: 'aiAssistant', label: 'KI-Assistent', description: 'Lokal zusammenfassen und Fragen stellen', icon: '🤖', requiresAi: true },
    ],
  },
  {
    key: 'office', title: 'Word & Office', icon: '🗃️',
    tools: [
      { id: 'wordToPdf', label: 'Word → PDF', description: 'DOCX zuverlässig konvertieren', icon: '📝' },
      { id: 'wordMerge', label: 'Word verbinden', description: 'Mehrere DOCX-Dateien zusammenführen', icon: '📋' },
      { id: 'wordDiff', label: 'Word-Vergleich', description: 'Inhaltliche Unterschiede sichtbar machen', icon: '🔍' },
      { id: 'wordMeta', label: 'Word-Metadaten', description: 'Dokumenteigenschaften bearbeiten', icon: 'ℹ️' },
      { id: 'excelMeta', label: 'Excel-Metadaten', description: 'Arbeitsmappeneigenschaften bearbeiten', icon: 'ℹ️' },
      { id: 'officeToPdf', label: 'Office → PDF', description: 'Excel, PowerPoint, ODF und RTF konvertieren', icon: '🗃️' },
    ],
  },
] as const satisfies readonly ToolGroupDefinition[]

export type ToolId = typeof TOOL_GROUPS[number]['tools'][number]['id']
export type ToolGroupKey = typeof TOOL_GROUPS[number]['key']
/** Die Schnittstellen-Intersection ergänzt optionale Merkmale, ohne die ID-Literale zu verlieren. */
export type CatalogTool = typeof TOOL_GROUPS[number]['tools'][number] & ToolDefinition
export type CatalogGroup = typeof TOOL_GROUPS[number]

export function findTool(toolId: string | null | undefined):
  | { tool: CatalogTool; group: CatalogGroup }
  | null {
  if (!toolId) return null
  for (const group of TOOL_GROUPS) {
    const tool = group.tools.find((entry) => entry.id === toolId)
    if (tool) return { tool, group }
  }
  return null
}
